File: robot-stream/output-webview/src/decoder.ts
// Based on the python decoder example.
// Used: https://extendsclass.com/python-to-javascript.html

import parseISO from "date-fns/parseISO";

export interface IMessage {
    message_type: string;
    decoded: any;
}

function version_decode(decoder, message) {
    return { message };
}

function simple_decode(decoder, message) {
    return JSON.parse(message);
}

function decode_time(decoder, time) {
    decoder.initial_time = parseISO(time);
    return { time };
}

function decode_memo(decoder, message) {
    var memo_id, memo_value;
    const split = splitInChar(message, ":");
    if (split) {
        [memo_id, memo_value] = split;
        try {
            memo_value = JSON.parse(memo_value);
        } catch (error) {
            console.log("Error parsing json: " + memo_value);
            console.log(error);
            return null;
        }
        decoder.memo[memo_id] = memo_value;
    }

    return null;
}

function start_suite(decoder, message) {
    let ident = decoder.ident;
    decoder.level += 1;
    let [name_id, suite_id_id, suite_source_id, time_delta_in_seconds] = message.split("|");
    let name = decoder.memo[name_id];
    let suite_id = decoder.memo[suite_id_id];
    let source = decoder.memo[suite_source_id];
    return { name, suite_id, source, time_delta_in_seconds };
}

function end_suite(decoder, message) {
    var ident, status, status_id, time_delta_in_seconds;
    decoder.level -= 1;
    ident = decoder.ident;
    [status_id, time_delta_in_seconds] = message.split("|");
    status = decoder.memo[status_id];
    return { status, time_delta_in_seconds };
}

function start_task_or_test(decoder, message) {
    let ident = decoder.ident;
    decoder.level += 1;
    let [name_id, suite_id_id, lineno, time_delta_in_seconds] = message.split("|");
    let name = decoder.memo[name_id];
    let suite_id = decoder.memo[suite_id_id];
    lineno = parseInt(lineno);
    return { name, suite_id, lineno, time_delta_in_seconds };
}

function end_task_or_test(decoder, message) {
    var ident, message_id, status, status_id, time_delta_in_seconds;
    decoder.level -= 1;
    ident = decoder.ident;
    [status_id, message_id, time_delta_in_seconds] = message.split("|");
    status = decoder.memo[status_id];
    message = decoder.memo[message_id];
    return { status, message, time_delta_in_seconds };
}

function _decodeOid(decoder, oid) {
    return decoder.memo[oid];
}
function _decodeFloat(decoder, msg) {
    return parseInt(msg);
}
function _decodeInt(decoder, msg) {
    return parseFloat(msg);
}

function _decode(message_definiton) {
    const names = [];
    const nameToDecode = new Map();
    for (let s of message_definiton.split(",")) {
        s = s.trim();
        const i = s.indexOf(":");
        let decode = "oid";
        if (i != -1) {
            [s, decode] = s.split(":");
        }
        names.push(s);
        if (decode === "oid") {
            nameToDecode.set(s, _decodeOid);
        } else if (decode === "int") {
            nameToDecode.set(s, _decodeInt);
        } else if (decode === "float") {
            nameToDecode.set(s, _decodeFloat);
        } else {
            throw new Error("Unexpected: " + decode);
        }
    }

    function _decImpl(decoder, message) {
        const splitted = message.split("|");
        const ret = {};
        for (let index = 0; index < splitted.length; index++) {
            const s = splitted[index];
            const name = names[index];
            ret[name] = nameToDecode.get(name)(decoder, s);
        }
        console.log('decoded', ret)
        return ret;
    }
    return _decImpl;
}

const start_keyword = _decode(
    "name:oid, libname:oid, keyword_type:oid, doc:oid, source:oid, lineno:int, time_delta_in_seconds:float"
);

// function start_keyword(decoder, message) {
//     let ident = decoder.ident;
//     decoder.level += 1;
//     let [name_id, libname_id, type_id, doc_id, source_id, lineno, time_delta_in_seconds] = message.split("|");
//     let keyword_type = decoder.memo[type_id];
//     let name = decoder.memo[name_id];
//     let libname = decoder.memo[libname_id];
//     let doc = decoder.memo[doc_id];
//     let source = decoder.memo[source_id];
//     lineno = parseInt(lineno);
//     return { keyword_type, name, libname, doc, time_delta_in_seconds, source, lineno };
// }

function end_keyword(decoder, message) {
    var ident, status, status_id, time_delta_in_seconds;
    decoder.level -= 1;
    ident = decoder.ident;
    [status_id, time_delta_in_seconds] = message.split("|");
    status = decoder.memo[status_id];
    return { status, time_delta_in_seconds };
}

function keyword_argument(decoder, message) {
    return decoder.memo[message];
}

const _MESSAGE_TYPE_INFO = {
    "V": version_decode,
    "I": simple_decode,
    "T": decode_time,
    "M": decode_memo,
    "SS": start_suite,
    "ES": end_suite,
    "ST": start_task_or_test,
    "ET": end_task_or_test,
    "SK": start_keyword,
    "EK": end_keyword,
    "KA": keyword_argument,
};

export class Decoder {
    memo;
    initial_time;
    level;
    ident;

    constructor() {
        this.memo = {};
        this.initial_time = null;
        this.level = 0;
        this.ident = "";
    }

    decode_message_type(message_type, message) {
        var handler;
        handler = _MESSAGE_TYPE_INFO[message_type];
        return handler(this, message);
    }
}

function splitInChar(line: string, char: string) {
    const i = line.indexOf(char);
    if (i > 0) {
        const message_type = line.substring(0, i);
        const message = line.substring(i + 1);
        return [message_type, message];
    }
    return undefined;
}

export function* iter_decoded_log_format(stream: string) {
    var decoded, decoder, message, message_type;
    decoder = new Decoder();

    for (let line of stream.split(/\r?\n/)) {
        line = line.trim();

        if (line) {
            const split = splitInChar(line, " ");
            if (split) {
                [message_type, message] = split;
                decoded = decoder.decode_message_type(message_type, message);

                if (decoded) {
                    const m: IMessage = { "message_type": message_type, "decoded": decoded };
                    yield m;
                }
            }
        }
    }
}

File: robot-stream/src/robot_stream/_decoder.py
import datetime
import json


def version_decode(decoder, message: str):
    return {"version": message}


def simple_decode(decoder, message: str):
    return {"info": json.loads(message)}


def decode_time(decoder, time):
    decoder.initial_time = datetime.datetime.fromisoformat(time)
    return {"initial_time": time}


def decode_memo(decoder, message):
    memo_id, memo_value = message.split(":", 1)
    memo_value = json.loads(memo_value)
    decoder.memo[memo_id] = memo_value
    return None


def _decode_oid(decoder, oid):
    return decoder.memo[oid]


def _decode_float(decoder, msg):
    return float(msg)


def _decode_int(decoder, msg):
    return int(msg)


def _decode_str(decoder, msg):
    return msg


red = "\u001b[31m"
green = "\u001b[32m"
yellow = "\u001b[33m"
blue = "\u001b[34m"
magenta = "\u001b[35m"
cyan = "\u001b[36m"
white = "\u001b[37m"
reset = "\u001b[0m"


def _color_status(status):
    if status == "PASS":
        return f"{green}{status}{reset}"
    elif status == "NOT RUN":
        return f"{cyan}{status}{reset}"
    else:
        return f"{red}{status}{reset}"


def _color_name(status):
    return f"{yellow}{status}{reset}"


def _decode(message_definition, level_diff=0):
    names = []
    name_to_decode = {}
    for s in message_definition.split(","):
        s = s.strip()
        i = s.find(":")
        decode = "oid"
        if i != -1:
            s, decode = s.split(":")
        names.append(s)
        if decode == "oid":
            name_to_decode[s] = _decode_oid

        elif decode == "int":
            name_to_decode[s] = _decode_int

        elif decode == "float":
            name_to_decode[s] = _decode_float

        elif decode == "str":
            name_to_decode[s] = _decode_str

        else:
            raise RuntimeError(f"Unexpected: {decode}")

    def dec_impl(decoder, message):
        decoder.level += level_diff
        splitted = message.split("|")
        ret = {}
        for i, s in enumerate(splitted):
            name = names[i]
            try:
                ret[name] = name_to_decode[name](decoder, s)
            except:
                ret[name] = None
        return ret

    return dec_impl


start_suite = _decode(
    "name:oid, suite_id:oid, suite_source:oid, time_delta_in_seconds:float",
    level_diff=+1,
)

end_suite = _decode("status:oid, time_delta_in_seconds:float", level_diff=-1)

start_task_or_test = _decode(
    "name:oid, suite_id:oid, lineno:int, time_delta_in_seconds:float", level_diff=+1
)

end_task_or_test = _decode(
    "status:oid, message:oid, time_delta_in_seconds:float", level_diff=-1
)

start_keyword = _decode(
    "name:oid, libname:oid, keyword_type:oid, doc:oid, source:oid, lineno:int, time_delta_in_seconds:float",
    level_diff=+1,
)

end_keyword = _decode("status:oid, time_delta_in_seconds:float", level_diff=-1)

decode_log = _decode("level:str, message:oid, time_delta_in_seconds:float")

assign = _decode("assign:oid")

tag = _decode("tag:oid")

keyword_argument = _decode("argument:oid")

start_time = _decode("start_time_delta:float")


_MESSAGE_TYPE_INFO = {
    "V": version_decode,
    "I": simple_decode,
    "T": decode_time,
    "M": decode_memo,
    "L": decode_log,
    "SS": start_suite,
    "ES": end_suite,
    "ST": start_task_or_test,
    "ET": end_task_or_test,
    "SK": start_keyword,
    "EK": end_keyword,
    "KA": keyword_argument,
    "AS": assign,
    "TG": tag,
    "S": start_time,
}


class Decoder:
    def __init__(self):
        self.memo = {}
        self.initial_time = None
        self.level = 0

    @property
    def ident(self):
        return "    " * self.level

    def decode_message_type(self, message_type, message):
        handler = _MESSAGE_TYPE_INFO[message_type]
        ret = {"message_type": message_type}
        try:
            r = handler(self, message)
            if not r:
                if message_type == "M":
                    return None
                raise RuntimeError(
                    f"No return when decoding: {message_type} - {message}"
                )
                if not isinstance(r, dict):
                    ret[
                        "error"
                    ] = f"Expected dict return when decoding: {message_type} - {message}. Found: {ret}"

            ret.update(r)
        except Exception as e:
            ret["error"] = f"Error decoding: {message_type}: {e}"
        return ret


def iter_decoded_log_format(stream):
    decoder = Decoder()
    for line in stream.readlines():
        line = line.strip()
        if line:
            message_type, message = line.split(" ", 1)
            decoded = decoder.decode_message_type(message_type, message)
            if decoded:
                yield decoded
